File: 2016-1-15/riddler.r
# https://fivethirtyeight.com/features/will-the-neurotic-basketball-player-make-his-next-free-throw/

library(pbapply)

shootFreethrows <- function(){
  shots <- c(1, 0)
  for(i in 3:100){
    p <- mean(shots)
    shot <- sample(x = c(1, 0), size = 1, prob = c(p, 1 - p))
    shots <- c(shots, shot)
  }
  if(shots[99] == 1){
    return(shots[100])
  }
}

shootFreethrowsRepeatedly <- function(){
  results <- unlist(pbreplicate(10000000, shootFreethrows()))
  return(mean(results))
}

File: 2016-1-15/riddler.py
# https://fivethirtyeight.com/features/will-the-neurotic-basketball-player-make-his-next-free-throw/

import random

def shootFreethrows():
    shots = [1,0]
    for i in range(2,100):
        p = sum(shots) / len(shots)
        shot = random.choices([1,0], weights = [p, 1 - p], k = 1)[0]
        shots.append(shot)
    if(shots[98] == 1):
        return shots[99]

def shootFreethrowsRepeatedly():
    results = []
    for i in range(10000000):
        result = shootFreethrows()
        if(result == 1 or result == 0):
            results.append(result)
    return sum(results) / len(results)

shootFreethrowsRepeatedly()
